test/gtest/ucp/ucp_test.h

```diff
@@ -212,6 +213,48 @@ class ucp_test : public ucp_test_base,
     volatile int m_err_handler_count;
     static const ucp_datatype_t DATATYPE;
     static const ucp_datatype_t DATATYPE_IOV;
+
+protected:
+    class mapped_buffer {
+    public:
+        mapped_buffer(size_t size, const entity& entity, int flags = 0,
+                      uint64_t seed = 0,
+                      ucs_memory_type_t mem_type = UCS_MEMORY_TYPE_HOST);
+        virtual ~mapped_buffer();
+
+        void *ptr() const;
+        uintptr_t addr() const;
+        size_t length() const;
+        ucs_memory_type_t mem_type() const;
+
+        void pattern_fill(uint64_t seed);
+        void pattern_check(uint64_t seed);
+        void pattern_fill();
+        void pattern_check();
+
+        ucp_rkey_h rkey(const entity& entity) const;
+        ucp_mem_h memh() const;
+
+        class ucp_rkey {
+        public:
+            ucp_rkey(ucp_rkey_h rkey) : m_rkey(rkey) {}
+            ~ucp_rkey() {ucp_rkey_destroy(m_rkey);}
+            operator ucp_rkey_h() {return m_rkey;}
+
+        private:
+            ucp_rkey_h m_rkey;
+        };
+
+    private:
+        const entity& m_entity;
+        mem_buffer    m_buffer;
+        uint64_t      m_seed;
+        ucp_mem_h     m_memh;
+        void         *m_rkey;
+        void         *m_ptr;
```

Comment:
can we inherit from mem_buffer instead of holding it as member, would not need to redefine some fields like ptr and size. Then, move the wrappers for pattern_fill etc. to ucs::mem_buffer base class.

Reply:
m_ptr & m_length could be filled/updated by mem_map call, it is filled by mem_query call.
didn't get comment about pattern* functions, removed it at all because unused for now

Reply:
you can add static function allocate(size, entity, flags, ...) and use to to call parent ctor from derived class ctor